### ISSUE TITLE
Unifi 2-way audio

### ISSUE THREAD

**Issue by victor-perez** [AlexxIT/go2rtc#279]
I'm trying to get the 2-way audio working on my Unifi doorbell while I can hear my voice on the streem it never comes out of the speaker of the doorbell. Did anybody manage to get 2-way audio working for Unifi?

**Comment by AlexxIT** [comment]
I have never heard that Unify uses open standards for two way audio.

**Comment by scottt732** [comment]
> AAC/ASTS is a problem. Browsers don't support receiving this codec from microphone. Maybe codec can be changed?

Just throwing this out there, but maybe wasm ([ffmpeg](https://chcunningham-samples.netlify.app/ffmpeg-wasm-demuxer/wasm_demuxer_demo.html)) and/or [WebCodecs API](https://developer.mozilla.org/en-US/docs/Web/API/WebCodecs_API) could be viable options? It looks like WebCodecs API is [available](https://caniuse.com/webcodecs) in recent Chrome/Edge builds.

**Comment by scelfo** [comment]
Could I help with this feature request somehow? I tried installing Scrypted and it is able to send two way audio from Chrome to a Unifi camera. I don't know whether it's doing server side transcoding or not.

I've been building out some nice HA dashboards with go2rtc and it would be awesome to have two way audio to my Unifi cameras via go2rtc.

**Comment by Glis6** [comment]
Also very interested in seeing this one. There seems to be no good solution for Unifi 2-way audio integration in Home Assistant at the moment.

**Comment by scelfo** [comment]
+1

I've tried several things since February and am now more confident that there is no good solution for Unifi 2-way audio.

**Comment by kksligh** [comment]
While I still use Frigate as my primary NVR - I have discovered that Scrypted has a plugin for Unifi cameras that allows consistent 2 way communication over HTTPS.  You have to use their Card and you have to pay for the basic NVR services to unlock the API token, but it works consistently.  

https://docs.scrypted.app/home-assistant.html

**Comment by Jens-Wymeersch** [comment]
I'm using Scrypted but the lag on some of the cameras is too long for me. Did anyone sort this out in the meanwhile ?

**Comment by tmryvz** [comment]
Has anyone new infos to this? I have 2way audio when I provide my unifi protect doorbell via homebridge to apple homekit. No problems there. But can't figure it out with go2rtc. Can anyone help please?

**Comment by Jens-Wymeersch** [comment]
Did you look at Scrypted ?

**Comment by Jens-Wymeersch** [comment]
The only way I found was using the Scrypted iframes.

**Comment by tmryvz** [comment]
Is this only for the paid scrypted nvr version?

**Comment by Jens-Wymeersch** [comment]
This is correct.

**Comment by tmryvz** [comment]
If only go2rtc would provide a working stream for 2-way audio. 
I want to use a open source and free method. 
Can you think of any other way to use doorbel g4 pro in Home Assistant for 2-way audio?

I mean it is possible. Homekit and Scrypted works fine...

**Comment by ViViDboarder** [comment]
Looks like HomeBridge uses ffmpeg to transcode the stream and passes it to the web socket. 

https://github.com/hjdhjd/homebridge-unifi-protect/blob/87fc75de4671ded6c0f5a7eb194c5721b8f35818/src/protect-stream.ts#L797

It stands to reason the same approach could be used. Some of the magic values could probably be copied from there.

**Comment by dudo** [comment]
I'm interested in this for the standalone docker image. I'll dig into the code a bit when I have time. Subscribed for now.

**Comment by t3hk0d3** [comment]
```
curl --http1.1 -v -k -X POST 'https://192.168.1.1/proxy/protect/integration/v1/cameras/681e457f03b8dc03e4256d97/talkback-session' -H "X-API-KEY: $UNIFI_API_KEY" -H 'Accept: application/json'| jq
> POST /proxy/protect/integration/v1/cameras/681e457f03b8dc03e4256d97/talkback-session HTTP/1.1
> Host: 192.168.1.1
> User-Agent: curl/8.10.1
> X-API-KEY: <redacted>
> Accept: application/json
> 
< HTTP/1.1 200 OK
< Server: nginx
< Date: Thu, 15 May 2025 11:54:00 GMT
< Content-Type: application/json; charset=utf-8
< Content-Length: 88
< Connection: keep-alive
< 
{
  "url": "rtp://192.168.1.56:7004",
  "codec": "opus",
  "samplingRate": 24000,
  "bitsPerSample": 16
}
```

Seems to be working. Looks like it always returns same URL/params.

**Comment by t3hk0d3** [comment]
🎉 

Seems like that URL is working any time, and i can send audio at any moment of time, and it is actually played:

```
ffmpeg -re -i sample-9s.wav -vcodec copy -vn -acodec libopus -f rtp rtp://192.168.1.56:7004
ffmpeg version n7.0.2 Copyright (c) 2000-2024 the FFmpeg developers
  built with gcc 14.2.1 (GCC) 20240910
  configuration: --prefix=/usr --disable-debug --disable-static --disable-stripping --enable-amf --enable-avisynth --enable-cuda-llvm --enable-lto --enable-fontconfig --enable-frei0r --enable-gmp --enable-gnutls --enable-gpl --enable-ladspa --enable-libaom --enable-libass --enable-libbluray --enable-libbs2b --enable-libdav1d --enable-libdrm --enable-libdvdnav --enable-libdvdread --enable-libfreetype --enable-libfribidi --enable-libgsm --enable-libharfbuzz --enable-libiec61883 --enable-libjack --enable-libjxl --enable-libmodplug --enable-libmp3lame --enable-libopencore_amrnb --enable-libopencore_amrwb --enable-libopenjpeg --enable-libopenmpt --enable-libopus --enable-libplacebo --enable-libpulse --enable-librav1e --enable-librsvg --enable-librubberband --enable-libsnappy --enable-libsoxr --enable-libspeex --enable-libsrt --enable-libssh --enable-libsvtav1 --enable-libtheora --enable-libv4l2 --enable-libvidstab --enable-libvmaf --enable-libvorbis --enable-libvpl --enable-libvpx --enable-libwebp --enable-libx264 --enable-libx265 --enable-libxcb --enable-libxml2 --enable-libxvid --enable-libzimg --enable-libzmq --enable-nvdec --enable-nvenc --enable-opencl --enable-opengl --enable-shared --enable-vapoursynth --enable-version3 --enable-vulkan
  libavutil      59.  8.100 / 59.  8.100
  libavcodec     61.  3.100 / 61.  3.100
  libavformat    61.  1.100 / 61.  1.100
  libavdevice    61.  1.100 / 61.  1.100
  libavfilter    10.  1.100 / 10.  1.100
  libswscale      8.  1.100 /  8.  1.100
  libswresample   5.  1.100 /  5.  1.100
  libpostproc    58.  1.100 / 58.  1.100
[aist#0:0/pcm_s16le @ 0x635d62ad2b40] Guessed Channel Layout: stereo
Input #0, wav, from 'sample-9s.wav':
  Duration: 00:00:09.59, bitrate: 1411 kb/s
  Stream #0:0: Audio: pcm_s16le ([1][0][0][0] / 0x0001), 44100 Hz, stereo, s16, 1411 kb/s
Stream mapping:
  Stream #0:0 -> #0:0 (pcm_s16le (native) -> opus (libopus))
Press [q] to stop, [?] for help
[libopus @ 0x635d62aa0e80] No bit rate set. Defaulting to 96000 bps.
Output #0, rtp, to 'rtp://192.168.1.56:7004':
  Metadata:
    encoder         : Lavf61.1.100
  Stream #0:0: Audio: opus, 48000 Hz, stereo, s16, 96 kb/s
      Metadata:
        encoder         : Lavc61.3.100 libopus
SDP:
v=0
o=- 0 0 IN IP4 127.0.0.1
s=No Name
c=IN IP4 192.168.1.56
t=0 0
a=tool:libavformat 61.1.100
m=audio 7004 RTP/AVP 97
b=AS:96
a=rtpmap:97 opus/48000/2
a=fmtp:97 sprop-stereo=1

[out#0/rtp @ 0x635d62ad2cc0] video:0KiB audio:117KiB subtitle:0KiB other streams:0KiB global headers:0KiB muxing overhead: 4.857637%
size=     123KiB time=00:00:09.58 bitrate= 104.8kbits/s speed=1.06x   
```

**Comment by t3hk0d3** [comment]
Well, i was wrong:
-  Just tried with another camera.
- You need to call that endpoint at least once to play audio thru RTP.
- Without calling endpoint you still can send audio, but it won't be played.
- After calling endpoint you can stream audio at any moment of time (it doesn't seems to expire even after 15 minutes).

**Comment by t3hk0d3** [comment]
I have feeling that two-way audio can be implemented with just super-tricky `ffmpeg_source`  config - we would need to somehow combine both ffmpeg RTSP input stream and  RTP audio output stream

**Comment by t3hk0d3** [comment]
![Image](https://github.com/user-attachments/assets/5a130bf5-c31d-482e-81d5-bfd4def092d4)

I wonder if `Stream #0:1: Audio: opus, 48000 Hz, stereo, fltp` is actually a talk-back channel 🤔

**Comment by t3hk0d3** [comment]
Somehow made it work with exec:

```
streams:
    unify-frakenstein: exec:ffmpeg -re -i rtsps://192.168.1.1:7441/aSrjYl35MXKCsF22 -i pipe: -map 0:0 -map 0:2 -c copy -rtsp_transport tcp -f rtsp {output} -map 1:0 -ar 24000 -vn -acodec libopus -ar 24000 -sample_fmt s16 -f rtp rtp://192.168.2.125:7004
```

**Comment by t3hk0d3** [comment]
After some testing seems like following config works much better (go2rtc automatically route streams itself):

```
streams:
  test: 
  - rtspx://192.168.1.1:7441/aSrjYl35MXKCsF22#backchannel=0
  - "exec:ffmpeg -re -f mulaw -ar 8000 -i - -vn -acodec libopus -ar 24000 -sample_fmt s16 -b:a 32k -f rtp rtp://192.168.1.50:7004#backchannel=1"
```

**Comment by unknownsolo** [comment]
You are on a roll today. Love it!

**Comment by t3hk0d3** [comment]
@AlexxIT can we have native support for backchannel RTP streams? 🙏🏻

**Comment by AlexxIT** [comment]
It would be hard to do blindly without having such a camera. But I'll add it to the tasks. Might have to wait months when I have free time to do it.

**Comment by t3hk0d3** [comment]
@AlexxIT AFAIK RTP works just over UDP, so its possible to use just something like `ffplay rtp://localhost:1234` to listen for audio from backchannel

**Comment by t3hk0d3** [comment]
Btw, i've tried to use an `#backchannel=1#audio=opus/24000/1` suffix to force audio format to OPUS to avoid unnecessary transcoding. Theoretically it should work (according to [here](https://github.com/AlexxIT/go2rtc/blob/master/pkg/pcm/backchannel.go#L22) and [here](https://github.com/AlexxIT/go2rtc/blob/master/internal/exec/exec.go#L89)). Unfortunately it did not.

probe result:
```
{
  "producers": [
    ...
    {
      "id": 2149371408,
      "format_name": "exec",
      "protocol": "pipe",
      "medias": [
        "audio, sendonly, PCMA/8000, L16"
      ],
      "senders": [
        {
          "id": 34,
          "codec": {
            "codec_name": "ANY",
            "codec_type": ""
          },
          "parent": 33
        }
      ]
    }
...
```

**Comment by t3hk0d3** [comment]
Just found out a fun fact: Home Assitant's official Unifi Protect integration can already play audio on cameras: 

https://github.com/uilibs/uiprotect/blob/main/src/uiprotect/stream.py#L149

**Comment by AlexxIT** [comment]
@t3hk0d3 are you sure you using latest master version? Because latest changes with exec backchannel only added to master version.

**Comment by scelfo** [comment]
@AlexxIT were you able to get the more native way you mentioned earlier to work? Very excited for this feature!!

**Comment by AlexxIT** [comment]
This is a low priority task for me. And it's important to remember that I don't have any Unifi equipment at all for any tests.

**Comment by t3hk0d3** [comment]
@scelfo After current master get released as a new version it would be possible to replicate an almost perfect behaviour using just `exec`.

**Comment by t3hk0d3** [comment]
@AlexxIT new version release would be awesome. Last version (1.9.9) was released quite a long time ago.

**Comment by AlexxIT** [comment]
There is no problem using the master version in any form, so as not to wait for the release:
https://github.com/AlexxIT/go2rtc?tab=readme-ov-file#go2rtc-dev-version

**Comment by t3hk0d3** [comment]
> There is no problem using the master version in any form, so as not to wait for the release:
> https://github.com/AlexxIT/go2rtc?tab=readme-ov-file#go2rtc-dev-version

Only problem for me with nightly versions is that they might work today, but suddenly became utterly broken tomorrow.

Pre-release or RC would be much appreciated.

**Comment by Jawomo** [comment]
Any progress here? Also very interested in this. Cant believe UI has no native solution for this...

**Comment by JustAnotherPU** [comment]
Any update to this ? Having a 2 way Audio working is a Dream

**Comment by spn91** [comment]
> it would be possible to replicate an almost perfect behaviour using just `exec`.

@t3hk0d3 I'm using the latest master version (as HA addon). Could you please provide the needed go2rtc config? Stream is working fine, but I don't know what you mean with "using just exec" to get twowayaudio working.

**Comment by digaus** [comment]
> Is this only for the paid scrypted nvr version?

You could iframe the scrypted page with some scripts which removes all other elements on load and auto starts the stream.

Could it working that way but it was tricky because you also need a permanent auth token to and create a session via websocket with it to be able to access it via iframe.

**Comment by binary-person** [comment]
Managed to get two-way audio working on G4 Instant in frigate v0.17.0-beta1 (go2rtc v1.9.10)

System versions:
- UCG Max: 4.4.9
- Protect: 6.2.72
- G4 Instant firmware: 5.1.217

## Step by step
(replace `192.168.1.1` with your protect server and `192.168.1.10` with IP of the unifi camera)

### get Protect api key and enable talkback on camera

- go to https://192.168.1.1/protect/settings/control-plane/integrations and generate your own API key
- Find the ID of your camera. should look something like `1588e101879356e8980cb3b7`
  - Use this: `curl -k -X GET 'https://192.168.1.1/proxy/protect/integration/v1/cameras' -H 'X-API-KEY: YOUR_API_KEY' -H 'Accept: application/json'`
- Enable talkback
  - `curl -k -X POST 'https://192.168.1.1/proxy/protect/integration/v1/cameras/1588e101879356e8980cb3b7/talkback-session' -H 'X-API-KEY: YOUR_API_KEY' -H 'Accept: application/json'`
  - response should look similar to this: `{"url":"rtp://192.168.1.10:7004","codec":"opus","samplingRate":24000,"bitsPerSample":16}`


### go2rtc config

```
# frigate users: replace ffmpeg with /usr/lib/ffmpeg/7.0/bin/ffmpeg
streams:
  G4_Instant:
    - rtspx://192.168.1.1:7441/aaaaaaaaaaaaaaaa#backchannel=0
    - exec:ffmpeg -re -fflags nobuffer -f alaw -ar 8000 -i - -vn -acodec libopus -b:a 32k -f rtp rtp://192.168.1.10:7004#backchannel=1#audio=pcma/8000
    - exec:ffmpeg -re -fflags nobuffer -f mulaw -ar 8000 -i - -vn -acodec libopus -b:a 32k -f rtp rtp://192.168.1.10:7004#backchannel=1#audio=pcmu/8000
    - exec:ffmpeg -re -fflags nobuffer -f s16le -ar 8000 -i - -vn -acodec libopus -b:a 32k -f rtp rtp://192.168.1.10:7004#backchannel=1#audio=pcml/8000
    - exec:ffmpeg -re -fflags nobuffer -f s16be -ar 8000 -i - -vn -acodec libopus -b:a 32k -f rtp rtp://192.168.1.10:7004#backchannel=1#audio=pcm/8000
```

Note for users on or below go2rtc v1.9.9: the audio specifier (e.g. `#audio=pcma/8000`) only works on or after v1.9.10. If the camera is outputting screeching noises, that's from codec mismatch and you're probably at or below v1.9.9. For a temporary fix, you can try checking which codec the browser defaulted to sending and move the corresponding codec to the top.

I had screeching noises coming out of examples provided by t3hk0d3 and it was because ffmpeg was encoding alaw as s16be. The audio parameter introduced in v1.9.10 should prevent this.

After this, it should work. If not, see below for my debugging methods

## debugging methods

- Send audio test file to camera:  `ffmpeg -re -i audio.wav -ar 24000 -vn -acodec libopus -payload_type 97 -b:a 32k -f rtp rtp://192.168.1.10:7004`
- Make sure camera is receiving RTP packets:
  - put the camera under different VLAN than your own current network, ssh into gateway, then run `tcpdump -i any -nn -vv -s 0 'udp and dst host 192.168.1.10 and dst port 7004'`
  - you can test tcpdump is working with push-to-talk on the browser or the unifi protect app


### create local RTP server to test go2rtc is sending audio correctly

Write the following to `recv.sdp`
```
v=0
o=- 0 0 IN IP4 0.0.0.0
s=RTP Opus
c=IN IP4 0.0.0.0
t=0 0
m=audio 7004 RTP/AVP 97
a=rtpmap:97 opus/24000/1
a=recvonly
```

Run this to listen for RTP on port 7004 udp and write to WAV: `ffmpeg -protocol_whitelist file,udp,rtp -i recv.sdp -c:a pcm_s16le out.wav`

Run this to test the RTP server is working (also works for Protect camera's RTP endpoint): `ffmpeg -re -i audio.wav -ar 24000 -vn -acodec libopus -payload_type 97 -b:a 32k -f rtp rtp://127.0.0.1:7004`


## other notes

- should be trivial to get browser to send OPUS directly to the camera. Going to wait until https://github.com/AlexxIT/go2rtc/pull/2019 gets merged and frigate picks up the version update
- not sure how long after hitting the api, talkback is turned back off. Maybe an idea for an official integration would be to hit this api before each talkback session
  - on this note, the api lets you find ID with camera mac, so the official integration could look something like this
  - `unifiprotect: 192.168.1.1/aaaaaaaa#mac=AABBCCDDEEFF#backchannel=1`
  - if someone figures out how to do it directly without protect server, we could have another plugin `unifiprotect_standalone`
- Thank you @t3hk0d3 for the clues to get it working

**Comment by t3hk0d3** [comment]
@binary-person great job!

once your fix is merged it should also remove double transcoding.

Currently there is transcoding (via ffmpeg) inside go2rtc to exec input format, and then another transcoding inside `exec` ffmpeg process to opus.

It ain't nice.
And if incoming connection is supporting OPUS too than this is completely waste of computing power and electricity.




However with native integration everything should be working much nicer, since ffmpeg is creating a huge latency/memory overhead. go2rtc internally can easily send RTP packets itself, with much smaller latency.

I ain't Go expert, but i have a feeling that having native RTP/Unifi integration (without FFMPEG) shouldn't be too much difficulty.

**Comment by jacobwtyler** [comment]
For anyone who may find this useful, I've solved two-way audio for Frigate with no echo.  PR submitted here: [PR: Fix echo feedback loop during two-way audio (talkback)](https://github.com/blakeblackshear/frigate/pull/22456)

**Comment by AlexxIT** [comment]
Implementing two-way communication for Unifi seems like such a simple task. It can be done in a single evening, with a few breaks. It's a shame that without a live camera and Unifi Protect, this is pretty difficult to implement.

**Comment by kksligh** [comment]
I’m happy to donate and ship a camera for testing if you’re open to that.

Thanks,

Kale Sligh
________________________________

**Comment by AlexxIT** [comment]
Unfortunately, it’s difficult to send anything to my address. All money transfers and shipments have been frozen for several years now.

**Comment by t3hk0d3** [comment]
If you can provide some ideas i'll try to vibe-code Unifi Protect support 😄

**Comment by kksligh** [comment]
Unfortunately I can’t help coding but I’m happy to help with funding equipment to test this functionality.  I use frigate a lot so I’d love to help here in any way

Thanks,

Kale Sligh
________________________________

**Comment by digaus** [comment]
> Implementing two-way communication for Unifi seems like such a simple task. It can be done in a single evening, with a few breaks. It's a shame that without a live camera and Unifi Protect, this is pretty difficult to implement.

Would be possible with giving you VPN access to a UCG Max running Protect with one camera attached wouldn't it?

**Comment by AlexxIT** [comment]
Implementing two-way audio via external access is a real headache. There's no feedback, and it's unclear whether it works or not.

**Comment by kksligh** [comment]
I appreciate everything you’re doing here - that said, this feature for unify for me is worth paying for.  I’m happy to compensate for your time and efforts to get this done.  I know this is a massive undertaking and understand if it’s just not worth it.  I wish I could contribute in other ways.

Thanks,

Kale Sligh
________________________________

**Comment by digaus** [comment]
> Implementing two-way audio via external access is a real headache. There's no feedback, and it's unclear whether it works or not.

Could work with two cameras 🙈

**Comment by starsoccer** [comment]
+1 would also gladly chip in to get this done

**Comment by flexyneat** [comment]
+1 too

**Comment by mpapierski** [comment]
Based on the discussion here, I opened a PR #2340 to add a `unifi-talkback:` channel to go2rtc. I'm using it in my Home Assistant, and it works really well in my setup. I tested it with a `Doorbell Lite` and `G4 Doorbell Pro`. There are example go2rtc configs there as well for testing
